ql/src/java/org/apache/hadoop/hive/ql/PreparePlanUtils.java

```diff
@@ -0,0 +1,263 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hadoop.hive.ql;
+
+import java.io.*;
+import java.util.*;
+
+import org.apache.hadoop.hive.common.type.*;
+import org.apache.hadoop.hive.conf.*;
+import org.apache.hadoop.hive.ql.exec.*;
+import org.apache.hadoop.hive.ql.exec.repl.util.ReplUtils;
+import org.apache.hadoop.hive.ql.exec.tez.*;
+import org.apache.hadoop.hive.ql.exec.vector.*;
+import org.apache.hadoop.hive.ql.hooks.HookUtils;
+import org.apache.hadoop.hive.ql.io.AcidUtils;
+import org.apache.hadoop.hive.ql.lockmgr.HiveTxnManager;
+import org.apache.hadoop.hive.ql.lockmgr.LockException;
+import org.apache.hadoop.hive.ql.log.PerfLogger;
+import org.apache.hadoop.hive.ql.metadata.AuthorizationException;
+import org.apache.hadoop.hive.ql.metadata.Hive;
+import org.apache.hadoop.hive.ql.metadata.HiveException;
+import org.apache.hadoop.hive.ql.optimizer.physical.*;
+import org.apache.hadoop.hive.ql.parse.*;
+import org.apache.hadoop.hive.ql.parse.type.*;
+import org.apache.hadoop.hive.ql.plan.*;
+import org.apache.hadoop.hive.ql.processors.CommandProcessorException;
+import org.apache.hadoop.hive.ql.security.authorization.command.CommandAuthorizer;
+import org.apache.hadoop.hive.ql.session.SessionState;
+import org.apache.hadoop.hive.ql.session.SessionState.LogHelper;
+import org.apache.hadoop.hive.ql.udf.generic.*;
+import org.apache.hadoop.hive.serde.*;
+import org.apache.hadoop.hive.serde2.typeinfo.*;
+import org.apache.hadoop.util.StringUtils;
+import org.apache.thrift.TException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.collect.ImmutableMap;
+
+/**
+ * The compiler compiles the command, by creating a QueryPlan from a String command.
+ * Also opens a transaction if necessary.
+ */
+public class PreparePlanUtils {
+
+  private static final Logger LOG = LoggerFactory.getLogger(PreparePlanUtils.class);
+
+  private static Set<Operator<?>> getAllFetchOperators(FetchTask task) {
+    if (task.getWork().getSource() == null)  {
+      return Collections.EMPTY_SET;
+    }
+    Set<Operator<?>> operatorList =  new HashSet<>();
+    operatorList.add(task.getWork().getSource());
+    return AnnotateRunTimeStatsOptimizer.getAllOperatorsForSimpleFetch(operatorList);
+  }
+
+  /*
+   * Retrieve name for PREPARE/EXECUTE statement
+   * Make sure the tree is either EXECUTE or PREPARE
+   */
+  protected static String getPrepareStatementName(ASTNode tree) {
+    if (tree.getType() == HiveParser.TOK_EXPLAIN) {
+      tree = (ASTNode)tree.getChildren().get(0);
+    }
+    assert (tree.getType() == HiveParser.TOK_PREPARE
+        || tree.getType() == HiveParser.TOK_EXECUTE);
+    return ((ASTNode)tree.getChildren().get(1)).getText();
+  }
+
+  public static void bindDynamicParams(QueryPlan plan, Map<Integer, String> parameterMap) {
```

Comment:
In order to avoid modifying the plan and adding per operator specific code below maybe we could use a UDF when generating the plan with dynamic parameters that performs a lookup in the map to obtain the correct value. Maybe other parts of the code could also be simplified like this.